tests/integration_base.go

```diff
@@ -76,14 +78,45 @@ type (
 		archivalNamespace      string
 		dynamicConfigOverrides map[dynamicconfig.Key]interface{}
 	}
+	// suiteParams contains the variables which are used to configure the test suite via the Option argument to
+	// setupSuite.
+	suiteParams struct {
+		fxOptions map[primitives.ServiceName][]fx.Option
+	}
+	Option func(params *suiteParams)
 )
 
-func (s *IntegrationBase) setupSuite(defaultClusterConfigFile string) {
+// WithFxOptionsForService returns an Option which, when passed as an argument to setupSuite, will append the given list
+// of fx options to the end of the arguments to the fx.New call for the given service. For example, if you want to
+// obtain the shard controller for the history service, you can do this:
+//
+//	var shardController shard.Controller
+//	s.setupSuite(t, tests.WithFxOptionsForService(primitives.HistoryService, fx.Populate(&shardController)))
+//	// now you can use shardController during your test
+//
+// This is similar to the pattern of plumbing dependencies through the TestClusterConfig, but it's much more convenient,
+// scalable and flexible. The reason we need to do this on a per-service basis is that there are separate fx apps for
+// each one.
```

Comment:
it would be provided to each one. I think it's probably better to add the qualifiers within the functions so that it's more opt-in. 

```go
type ServiceIndex int

// in the loop to provide multiple nodes:
fx.New(
  fx.Provide(func() ServiceIndex { return nodeIndex }),
  ...
)

// in the place where you provide the options
var someDep SomeType
WithFxOptionsForService(primitives.WorkerService, fx.Invoke(func(idx ServiceIndex, dep SomeType) {
  if idx == 0 {
    someDep = dep
  }
}))
```